[Table] New Table component

akhenry opened the pull request:
__DO NOT MERGE UNTIL__ #651 is merged as it is dependent upon it for regions. Indeed the build will fail until regions are merged.

Implements #670 . Merge of historical data functionality. #707 has been created to address realtime. 

### Summary of changes
- Integrated MctDataTable directive, refactored and renamed to MCTTable
- Created new object type 'Table'
- Created new view 'Table', with controller and markup
- Migrated some existing tabular code from the scrolling bundle across to the Table bundle.
- Created a new edit region in the inspector for customizing column visibility in table view.

larkin commented on platform/features/table/bundle.js:
```diff
@@ -0,0 +1,132 @@
+                    "depends": ["$scope"]
+                },
+                {
+                    "key": "MCTTableController",
```
One problem to registering this controller is it takes up namespace and is not reusable outside of the directive.  Would recommend loading this in the directive via require and specifying dependencies with the `$inject` property.  Keep the TableDirective as a single component that only exposes a single interface.

VWoeltjen replied:
:+1:, but for consistency I would steer away from `$inject` and specify `controller: ["$scope", "$timeout", "$element", MCTTableController]` in `mct-table`'s directive definition object.

akhenry replied:
Ah, yep, agreed, good suggestion.

larkin commented on platform/features/table/bundle.js:
```diff
@@ -0,0 +1,132 @@
+            ],
+            "controllers": [
+                {
+                    "key": "TableController",
```
I think a better name for this would be a TelemetryTableController: This controller deals with Telemetry and Tables.

akhenry replied:
Fair enough, I've renamed it.

larkin commented on platform/features/table/src/Table.js:
```diff
@@ -0,0 +1,163 @@
+                if (columnValue !== undefined && columnValue.text === undefined){
+                    columnValue.text = '';
+                }
+                // Don't replace something with nothing.
```
It seems that multiple columns with the same column title should be combined, and no datum should provide two different values for the same key.  How does this occur?

akhenry replied:
Yeah, this is kind of a subtle problem that I encountered. I decided that when dealing with tables composed of multiple telemetry points, domains and ranges with the same title should be shown in the same column. At the moment it is just based on Title, but I think that in future we could be a bit cleverer with this and only combine columns with the same title AND the same type [range or domain] AND the same formatting... or something. It's either that or don't combine them at all for now and just suffix them with the object name or something to distinguish them when they're displayed.

The reason for this logic is that you can have two telemetry points with domains or ranges that have the same Title (eg. Time) but different keys (time vs. timestamp). The data passed to the directive is keyed on title (because that's what is shown in the headers). So this can cause data loss if the the value for 'time' is overwritten by the value for 'timestamp' (which is undefined).

larkin commented on platform/features/table/src/controllers/TableController.js:
```diff
@@ -0,0 +1,214 @@
+            var i;
+            //this.pending = false;
+            for (i=0; i < series.getPointCount(); i++) {
+                this.updateRows(domainObject, this.handle.makeDatum(domainObject, series, i));
```
Okay, so some interesting thoughts on telemetry handle, datums, and limits.  First off, I'm not a big fan of telemetry handles/handlers/subscriptions-- would prefer we talk directly to the telemetry API (general idea: less interfaces).   That's my biggest comment here: what does it take to remove telemetry handlers from this code?

On the VISTA side, the TelemetrySeries implements a `getDatum` method which can be used in place of `makeDatum` with minor differences-- it does not parse the underlying values.  In VISTA, the datum values used for limits evaluation are not ranges or domains, so they won't come through if you use `makeDatum`, and alarm validation won't work.  This is a difference we need to figure out how to resolve.  I'd push for adding a `getDatum` method to all telemetry series to use instead of `makeDatum`.

`updateRows` could be changed to take a domainObject and a telemetrySeries.  From there, it should be able to pull out only the range/domain values it needs (based on column configuration)-- and not the entire datum.  Additionally, it seems wasteful to re-evaluate limits for each cell in a row, when the logic is essentially the same.  Can we assign the CSS class to the entire row and not the individual cells?

akhenry replied:
Let me do some analysis and see what it would take to pull out the telemetryHandler stuff

akhenry replied:
> Additionally, it seems wasteful to re-evaluate limits for each cell in a row, when the logic is essentially the same. Can we assign the CSS class to the entire row and not the individual cells?

The reasoning behind evaluating for each __range__ cell (doesn't do it for domains) is that with multiple range values, only one of them may be in limit violation, and it's useful to highlight which one it is.

larkin replied:
You're right, forgot that different ranges may have different criteria.

akhenry replied:
Ok, I've got a sense now for what's necessary to refactor this to remove telemetry handle/handler/subscriptions. It wouldn't be too painful, but I kind of feel as though it's a nice to have at this point. It's working with the existing API, and any additional effort invested now might be wasted in light of #463 and the plan to redesign the telemetry API. Any changes made to telemetry series objects also would need to be applied in WARP.

larkin replied:
One problem with this is that limit checking won't work in VISTA due to current implementations.  We can remedy that with fixes on the VISTA side or on the OMCT side.  In this case, it is probably  better to get this merged and start ensuring this integrates with VISTA.  

I'll log a ticket VISTA side to ensure this works properly.

larkin commented on platform/features/table/src/Table.js:
```diff
@@ -0,0 +1,163 @@
+         * @param domainObject
+         * @constructor
+         */
+        function Table(domainObject, telemetryFormatter) {
```
I think a more clear name for this class would be `TableConfiguration`: Given a domainObject, it tells you the table configuration.  The configuration handles formatting of values, too.

Interestingly, this class knows how to load configuration from a domain object yet doesn't know how to save configuration to a domain object (the TableController handles saving).  If it's going to load configuration, it should be able to save it too-- remove that concern from the TableController.

akhenry replied:
re: TableConfiguration. I'm a little concerned that it overloads the term 'configuration'. But I agree 'Table' doesn't give you much of a clue about what it does.

I debated where responsibility for object model configuration should reside. To be honest I'm not sure the Table class should be concerned with object model configuration at all. Having a concept of 'column configuration' is fine, but it might be better if it's agnostic of the source of that configuration, and just have it handed to it. That said, it simplifies the controller if the Table class handles both reading and writing model configuration, and simplification was the original motivation for creating the Table class at all.

larkin commented on platform/features/table/src/Table.js:
```diff
@@ -0,0 +1,163 @@
+    [
+        './DomainColumn',
+        './RangeColumn',
+        './NameColumn'
```
NameColumn is unused in this file-- TableController handles adding name columns.

It seems to me that name column should be handled by the TableConfiguration and not by the TableController.

akhenry replied:
In an earlier incarnation it was handled by Table/TableConfiguration. But the behavior of automatically adding a name column if there is more than one domain object is 'cleverness' that I figured was better left up to the calling class rather than baking it into the Table class.

akhenry replied:
Moreover, the Table class doesn't actually know the names of composed objects, it just handles their telemetry metadata.

akhenry replied:
>~~Moreover, the Table class doesn't actually know the names of composed objects, it just handles their telemetry metadata.~~ 

I take it back, NameColumn gets the name itself from the domainObject so it wouldn't be necessary to provide anything else to the Table class itself.

larkin commented:
@akhenry in general there is a mix of concerns between the TableController (which I would call the TelemetryTableController) and the Table class (which I would call TableConfiguration).  I think this can be much clearer: the table controller instantiates TableConfiguration, requests telemetry data, and provides all that to the view (the table directive).  Sending back your way.

akhenry commented:
> in general there is a mix of concerns between the TableController (which I would call the TelemetryTableController) and the Table class (which I would call TableConfiguration)

It's a fair point, and I had similar concerns myself. I'll address each of the points you've made, and try and clarify the distinction a little in the code.

akhenry commented:
### Changes
* Renamed Table (and TableSpec) to TableConfiguration(Spec)
* Renamed TableController to TelemetryTableController
* Removed MctTableController from global controllers namespace
* Moved addition of NameColumn into TableConfiguration class
* renamed some functions to make their role a little clearer.
* Moved saveColumnConfiguration onto the TableConfiguration class

Have deferred suggested changes to telemetry handling, see https://github.com/nasa/openmctweb/pull/708#discussion_r54837924

larkin commented:
@akhenry ready for re-review?  send back my way if so.

akhenry commented:
Done.

VWoeltjen commented:
FWIW, if we're trying to get this in before today's testathon, I think there's not much value there - current table is historical-only, testathon data will be realtime-only.

larkin commented:
@VWoeltjen I plan to run testathon against VISTA too to make sure we are catching any potential bugs there, so _some_ value in getting it in.

akhenry commented:
@larkin we're good to go.

VWoeltjen commented:
Even better!

larkin commented:
@akhenry In some testing locally, I have noticed that changes to table configuration while in edit mode trigger a persistence operation to the server, which generally fails with a new object because the object has not yet been created.  Doesn't seem to occur after the object has been saved and then edited a second time.  Not actually a blocker.

larkin commented:
Disregard my previous comment; that was the search indexer attempting to fetch the object from persistence, it was not a persistence call.  That is a known issue.

# Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y

akhenry commented:
@larkin Ah, #377 , that old chestnut. Oddly I didn't see this in my own testing, but it totally makes sense.
